File: TourismAgency/src/main/java/repository/database/AgentDbRepository.java
package repository.database;

import domain.TravelAgent;
import domain.validators.ValidationException;
import domain.validators.Validator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.AgentRepository;
import repository.RepoException;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class AgentDbRepository implements AgentRepository {

    private final JdbcUtils dbUtils;
    private final Validator<TravelAgent> agentValidator;

    private static final Logger logger= LogManager.getLogger();

    public AgentDbRepository(Properties props, Validator<TravelAgent> agentValidator){
        this.agentValidator = agentValidator;
        logger.info("Initializing AgentRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
    }
    @Override
    public TravelAgent findOne(Long aLong) {
        logger.traceEntry("finding agent with id {} ",aLong);
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select * from login where idf=?")){
            preStmt.setLong(1,aLong);
            try(ResultSet result=preStmt.executeQuery()) {
                if (result.next()) {
                    long id = result.getLong("idt");
                    String username = result.getString("username");
                    String passwd = result.getString("passw");
                    TravelAgent travelAgent=new TravelAgent(username,passwd);
                    travelAgent.setId(id);
                    logger.traceExit(travelAgent);
                    return travelAgent;
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("No travel agent found with id {}", aLong);

        return null;
    }

    @Override
    public Iterable<TravelAgent> findAll() {
        Connection con=dbUtils.getConnection();
        List<TravelAgent> agents=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from login")) {
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("idf");
                    String username = result.getString("username");
                    String passwd = result.getString("passw");
                    TravelAgent travelAgent=new TravelAgent(username,passwd);
                    travelAgent.setId(id);
                    agents.add(travelAgent);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(agents);
        return agents;
    }

    @Override
    public TravelAgent save(TravelAgent entity) throws ValidationException, RepoException {
        if (entity == null)
            throw new RepoException("Repository exception: id must be not null!\n");
        agentValidator.validate(entity);
        if (findOne(entity.getId()) != null) {
            return entity;
        } else{
            logger.traceEntry("saving agent {} ",entity);
            Connection con=dbUtils.getConnection();
            try(PreparedStatement preStmt=con.prepareStatement("insert into login(username,passw) values (?,?)")){
                //preStmt.setDouble(1,entity.getId());
                preStmt.setString(1,entity.getUsername());
                preStmt.setString(2,entity.getPasswd());

                int result=preStmt.executeUpdate();
                logger.traceEntry("rows affected {}",result);
            }catch (SQLException ex){
                logger.error(ex);
                System.out.println("Error DB "+ex);
            }
            logger.traceExit();
        }
        return null;
    }

    @Override
    public TravelAgent delete(Long aLong) throws RepoException {
        if (aLong == null)
            throw new RepoException("Repository exception: id must be not null!\n");

        logger.traceEntry("deleting travel agent with {}",aLong);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("delete from login where idf=?")){
            preStmt.setLong(1,aLong);
            int result=preStmt.executeUpdate();
            logger.traceEntry("rows affected {}",result);
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public TravelAgent update(TravelAgent entity) throws RepoException, ValidationException {
        if (entity == null)
            throw new RepoException("Repository exception: id must be not null!\n");
        agentValidator.validate(entity);
        if (findOne(entity.getId()) == null) {
            return entity;
        } else{
            logger.traceEntry("updating agent {} ",entity);
            Connection con=dbUtils.getConnection();
            try(PreparedStatement preStmt=con.prepareStatement("update login set username=?,passw=? where idf=?")){
                //preStmt.setDouble(1,entity.getId());
                preStmt.setString(1,entity.getUsername());
                preStmt.setString(2,entity.getPasswd());
                preStmt.setDouble(3,entity.getId());

                int result=preStmt.executeUpdate();
                logger.traceEntry("rows affected {}",result);
            }catch (SQLException ex){
                logger.error(ex);
                System.out.println("Error DB "+ex);
            }
            logger.traceExit();
        }
        return null;
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as [SIZE] from login")) {
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        return 0;
    }
}

File: AgentieTurism/persistence/AgentDBRepository.cs
using model.domain;
using System;
using System.Collections.Generic;

using System.Data;
using log4net;


namespace persistence
{


    public class AgentDBRepository : AgentRepository
    {
        private static readonly ILog log = LogManager.GetLogger("AgentDbRepository");

        public AgentDBRepository()
        {
            log.Info("Creating AgentDBRepository");
        }

        public void delete(long id)
        {
            log.InfoFormat("Deleting travel agent with id {0}",id);
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from login where idf=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                var dataR = comm.ExecuteNonQuery();

                if (dataR == 0)
                    throw new RepositoryException("No agent deleted!");
                else
                    log.Info("Travel agent deleted succesfully");
            }
        }

        

        public TravelAgent findAgentsByUser(string user, string passwd)
        {
            log.InfoFormat("Entering findOne with username: {0}, password: {1}", user,passwd);
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select idf,username, passw from login where username=@user and passw=@passwd";
                IDbDataParameter paramUser = comm.CreateParameter();
                paramUser.ParameterName = "@user";
                paramUser.Value = user;

                comm.Parameters.Add(paramUser);
                var paramPas = comm.CreateParameter();
                paramPas.ParameterName = "@passwd";
                paramPas.Value = passwd;
                comm.Parameters.Add(paramPas);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long idf = dataR.GetInt64(0);
                        String username = dataR.GetString(1);
                        String passw = dataR.GetString(2);
                        TravelAgent agent = new TravelAgent(username, passw);
                        agent.ID = idf;
                        log.InfoFormat("Exiting findOne with value {0}", agent);
                        return agent;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public IEnumerable<TravelAgent> findAll()
        {
            IDbConnection con = DBUtils.getConnection();
            IList<TravelAgent> agents = new List<TravelAgent>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select idf,username, passw from login";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long idf = dataR.GetInt64(0);
                        String username = dataR.GetString(1);
                        String passw = dataR.GetString(2);
                        TravelAgent agent = new TravelAgent(username, passw);
                        agent.ID = idf;
                        agents.Add(agent);
                    }
                }
            }

            return agents;
        }

        public TravelAgent findOne(long id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select idf,username, passw from login where idf=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long idf = dataR.GetInt64(0);
                        String username = dataR.GetString(1);
                        String passw = dataR.GetString(2);
                        TravelAgent agent = new TravelAgent(username, passw);
                        agent.ID = idf;
                        return agent;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void save(TravelAgent entity)
        {
            var con = DBUtils.getConnection();
            log.InfoFormat("Saving agent = {0}", entity);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into login (username,passw) values (@username, @passwd);select LAST_INSERT_ID()";
                /*var paramId = comm.CreateParameter();
                paramId.ParameterName = "@idf";
                paramId.Value = entity.ID;
                comm.Parameters.Add(paramId);*/

                var paramUser = comm.CreateParameter();
                paramUser.ParameterName = "@username";
                paramUser.Value = entity.Username;
                comm.Parameters.Add(paramUser);

                var paramPas = comm.CreateParameter();
                paramPas.ParameterName = "@passwd";
                paramPas.Value = entity.Passwd;
                comm.Parameters.Add(paramPas);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long idf = dataR.GetInt64(0);
                        
                        Console.WriteLine("LAST ID INSERTED: {0}", idf);
                        entity.ID = idf;
                        log.InfoFormat("Agent was added succesfully}");
                    }
                }
                
            }
        }

        public void update(TravelAgent entity)
        {
            var con = DBUtils.getConnection();
            log.InfoFormat("Updating agent = {0}", entity);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update login set username=@username ,passw=@passwd where idf=@idf";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@idf";
                paramId.Value = entity.ID;
                comm.Parameters.Add(paramId);

                var paramUser = comm.CreateParameter();
                paramUser.ParameterName = "@username";
                paramUser.Value = entity.Username;
                comm.Parameters.Add(paramUser);

                var paramPas = comm.CreateParameter();
                paramPas.ParameterName = "@passwd";
                paramPas.Value = entity.Passwd;
                comm.Parameters.Add(paramPas);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("Update failed !");
                else {
                    log.InfoFormat("Agent was updated succesfully!");
                }
            }
        }
    }
}
